packages/api/tests/lib/test-granules.js

```diff
@@ -91,3 +282,221 @@ test('getGranuleProductVolume() returns correct product volume', (t) => {
+  // prevent the granule from being deleted if files referencing it
+  // still exist.
+  /* eslint-disable no-await-in-loop */
+  for (let i = 0; i < files.length; i += 1) {
+    const file = files[i];
+    t.false(await s3ObjectExists({ Bucket: file.bucket, Key: file.key }));
+  }
+  /* eslint-enable no-await-in-loop */
+});
+
+test.serial('deleteGranuleAndFiles() succeeds if a file is not present in S3', async (t) => {
+  const granuleId = randomId('granule');
+  const files = [
+    {
+      bucket: process.env.system_bucket,
+      fileName: `${granuleId}.hdf`,
+      key: randomString(),
+    },
+  ];
+
+  // Create Dynamo granule
+  const newGranule = fakeGranuleFactoryV2({ granuleId: granuleId, status: 'failed' });
+  newGranule.published = false;
+  newGranule.files = files;
+  await granuleModel.create(newGranule);
+
+  // create PG granule
+  const fakePGGranule = fakeGranuleRecordFactory(
+    {
+      granule_id: granuleId,
+      status: 'failed',
+      collection_cumulus_id: t.context.testPgCollection.cumulus_id,
+    }
+  );
+  fakePGGranule.published = false;
+  const [granuleCumulusId] = await granulePgModel.create(t.context.knex, fakePGGranule);
+
+  const newPgGranule = await granulePgModel.get(t.context.knex, { cumulus_id: granuleCumulusId });
+  const newDynamoGranule = await granuleModel.get({ granuleId: newGranule.granuleId });
+
+  await deleteGranuleAndFiles({
+    knex: t.context.knex,
+    dynamoGranule: newDynamoGranule,
+    pgGranule: newPgGranule,
+    granuleModelClient: granuleModel,
+  });
+
+  // Check Dynamo and RDS. The granule should have been removed from both.
+  await t.throwsAsync(
+    granuleModel.get({ granuleId: newDynamoGranule.granuleId }),
+    { instanceOf: RecordDoesNotExist }
+  );
+
+  await t.throwsAsync(
+    granulePgModel.get(t.context.knex, { granule_id: newPgGranule.granule_id }),
+    { instanceOf: RecordDoesNotExist }
+  );
+});
+
+test.serial('deleteGranuleAndFiles() will not delete a granule or its S3 files if the PG file delete fails', async (t) => {
+  const { newPgGranule, newDynamoGranule, files } = await createGranuleAndFiles(
+    t.context.knex,
+    t.context.testPgCollection.cumulus_id,
+    false
+  );
+
+  const mockFileModel = {
+    tableName: 'files',
+    delete: () => {
+      throw new Error('Delete failed');
+    },
+  };
+
+  await t.throwsAsync(
+    deleteGranuleAndFiles({
+      knex: t.context.knex,
+      dynamoGranule: newDynamoGranule,
+      pgGranule: newPgGranule,
+      filePgModel: mockFileModel,
+      granuleModelClient: granuleModel,
+    }),
+    { instanceOf: Error }
+  );
+
+  // granule should still exist in Dynamo and PG
+  t.true(await granulePgModel.exists(t.context.knex, { granule_id: newPgGranule.granule_id }));
+  t.true(await granuleModel.exists({ granuleId: newDynamoGranule.granuleId }));
+
+  /* eslint-disable no-await-in-loop */
+  for (let i = 0; i < files.length; i += 1) {
```

Comment:
I prefer this to disabling ESlint throughout this file:

`files.forEach(async (file) => {`

Reply:
Yeah so I didn't use `forEach` because it's based on callbacks and that presents a problem when we have promises returning in those callbacks. Even if we `await` them, the `forEach` method will return immediately so the tests are never actually evaluated.

The other, more ES6-y way to do this would be to await `Promise.all` and use `map` I think. That's what I've done here:

26afa2ab08540d84652935532847a8a338f499e2

Note that there's actually one there that I originally had as a `forEach`. Whoops. I tried changing the expected return from `true` to `false` and the test still passed because that was never evaluated.